yt_idefix/data_structures.py

```diff
@@ -430,6 +458,260 @@ def _get_code_version(self) -> str:
+                if log_match:
+                    self.current_time = float(log_match.group(1))
+                    break
+            else:
+                ytLogger.warning(
+                    "Failed to retrieve time from %s, setting current_time = -1",
+                    out_file,
+                )
+
+    def _parse_definitions_header(self) -> None:
+        """Read some metadata from header file 'definitions.h'."""
+        self.parameters["definitions"] = {}
+        if not self._definitions_header:
+            ytLogger.warning(
+                "%s was not found. Code units will be set to 1.0 in cgs.",
+                self._default_definitions_header,
+            )
+            return
+
+        with open(self._definitions_header) as fh:
+            body = fh.read()
+        lines = C_io.strip_comments(body).split("\n")
+
+        for line in lines:
+            if (geom_match := re.fullmatch(_DEF_GEOMETRY_REGEXP, line)) is not None:
+                self.parameters["definitions"]["geometry"] = geom_match.group(1).lower()
+            elif (unit_match := re.fullmatch(_DEF_UNIT_REGEXP, line)) is not None:
+                unit = unit_match.group(1).lower() + "_unit"
+                expr = unit_match.group(2)
+                # Before evaluating the expression, replace the input parameters,
+                # pre-defined constants, code units and arithmetic operators
+                # that cannot be resolved. The order doesn't matter.
+                expr = re.sub(r"g_inputParam\[(\w+)\]", self._get_input_parameter, expr)
+                expr = re.sub(r"CONST_\w+", self._get_constants, expr)
+                expr = re.sub(r"UNIT_(\w+)", self._get_unit, expr)
+                expr = re.sub(r"sqrt", "np.sqrt", expr)
+                expr = re.sub(r"log", "np.log", expr)
+                self.parameters["definitions"][unit] = eval(expr)
+
+    def _get_input_parameter(self, match: re.Match) -> str:
+        """Replace matched input parameters with its value"""
+        key = match.group(1)
+        return str(self.parameters["Parameters"][key])
+
+    def _get_unit(self, match: re.Match) -> str:
+        """Replace matched unit with its value"""
+        key = match.group(1).lower() + "_unit"
+        return str(self.parameters["definitions"].get(key, 1.0))
+
+    def _get_constants(self, match: re.Match) -> str:
+        """Replace matched constant string with its value"""
+        key = match.group()
+        return str(pluto_def_constants[key])
+
+    def _get_code_version(self) -> str:
+        # take the first line of the header
+        # Code version is read from grid.out
+        """
+        In grid.out
+        # ***********
+        marks the beginning and the end of header information
+        """
+        if not (hasattr(self, "grid_file")):
+            self.grid_file = os.path.join(
+                self.directory, "grid.out"
+            )  # data-loc/grid.out
```

Comment:
It can be hard to refactor so I'm hesitant to make this comment but `hasattr` should be avoided as much as possible because it makes code much harder to reason about. Its usage may be justified in contexts where maintaining backward compatibility is important, but this is brand new code. There's a lot on your plate already so I can't *demand* that this be refactored out, but please avoid spreading this pattern as I will eventually need to get it of it.